### PR TITLE
Optimize seeds and mpi trans

### PR DESCRIPTION
**Pull request by markusbattarbee**
Contains the following improvements:
 * Only communicate VDFs for those spatial cells which are required, when doing AMR translation. (E.g. with a VLASOV_STENCIL_WIDTH of 2, we use a neighborhood of 3, which can include up to 24 cells, when we only need up to 8).
 * Calculate pencils only at load balance. Save each direction of pencils at that time.
 * Optimize selection of seed-cellids for pencils in order to execute the minimum required amount of Lagrangian remapping

Supercedes #539 

Testpackage is clean
[log_OptimizeSeedsAndMPITrans_tp.txt](https://github.com/fmihpc/vlasiator/files/6082460/log_OptimizeSeedsAndMPITrans_tp.txt)

_Improvement using a 3D two-AMR-level IPshock run (done for the AMR 6D method paper):_
**spatial space**
long 9.112e+06 Cells/s
MPI 9.883e+06 Cells/s 
seeds 8.578e+06 Cells/s 
both 1.006e+07 Cells/s

AMR_comparison_extra

Interesting that just seed-ids is actually slower, but combining them is faster. 

TBD: performance analysis vs thread counts at scale

### CODE REVIEW

**Comment by markusbattarbee** [comment]
Yann's performance tests on Hawk show it's consistently faster:
image

Also, now made long pencils default and moved the input parameter to the AMR group.

**Comment by ykempf** [comment]
Clean diffs, better performance, what is there not to love?

**Comment by ykempf** [comment]
Do we need the new DRO though?

**Comment by markusbattarbee** [comment]
Could be useful for visualization of MPI load in 3D runs? :) And it doesn't do any harm.

**Comment by markusbattarbee** [comment]
Performance after the skip empty vectors enhancement:

image

And the testpackage log:

[log.txt](https://github.com/fmihpc/vlasiator/files/6122825/log.txt)

**Comment by markusbattarbee** [comment]
So any comments on this, do we want to merge into dev (in preparation for 5.2) or fine-tune something more?

**Comment by ykempf** [comment]
The testpackage log looked photoshopped until I saw the last test. ;) Good to go in my opinion, the clean diffs and performance boost speak for themselves.

**Review by ykempf — APPROVED** [review]
And now as a formal review. Go!

**Comment by markusbattarbee** [comment]
Fixing this one seedid selection bug seems to also give a nice increase:
before: 9.594e+06 Cells/s  
after:  9.931e+06 Cells/s
Will confirm once test run has completed.

File: vlasovsolver/cpu_trans_map_amr.cpp [hunk 1]
```diff
@@ -716,37 +830,129 @@ void getSeedIds(const dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry>& mpiGr
...
+      /* -----------------------------------------
+         | A |   | B |   |_|_|_|_|   |   | C |   |
+         |   |   |   |   | | | | |   |   |   |   |
+         -----------------------------------------
+         For optimal pencil generation, we need seedids at A, B, and C.
+         A Is triggered in the first if-clause. Pencils starting from B
+         will be split (but won't cause A to split), and pencils from
+         C will be able to remain again un-split. These checks need to be done
+         only if we aren't already at the maximum refinement level.
+      */
```
Comment (ursg) [inline vlasovsolver/cpu_trans_map_amr.cpp:858]: This is awesome. Really good optimization, and elegant in its implementation, too!

**Review by ursg — APPROVED** [review]
Yep, I like this alot. The code is much more readable (on a conceptual level) than before, too.
And the performance results speak for themselves.

**Comment by markusbattarbee** [comment]
Run completed
after: 1.021e+07 Cells/s
this version is the AMRbothblocks2 bar, so we're close to per-cell performance of the old 2D approach now!
AMR_comparison_extra

**Comment by ursg** [comment]
Alright, into the dev branch you go!

**Pull request merged**